### PR TITLE
Suspected fix for progressive slowdown when training

### PR DESCRIPTION
**Pull request by Loca123**
When training on multiple GPUs, with a validation dataset, I have been seeing a significant reduction in training speed after each epoch. When validation epochs are sufficiently frequent, this is very detremental. This does not present when running on a single GPU with the validation set, or when running on multiple GPUs without the validation set.

The graph below shows a toy example to exascerbate the problem for illustration, but this occurs in a much larger network that was taking around a week to train on a single GPU. After 24 hours, with a high validation frequency, the system had effectively stopped training. I've spent several days logging times and plotting graphs on different hardware, operating systems, versions of tensorflow and with different networks and datasets and this is what I have narrowed it down to. It also appears to have an exponential nature, initially it can look like quite a steady increase in training time, but some of the tests that I ran showed, a rapid increase in time beyond a certain point.

image

Here's another graph. This shows Ubuntu with 4xGPUs on Amazon, 1 GPU on a local Windows machine and 2 GPUs on a local WIndows machine. Running TF 1.14 installed from pip. The issue is present in TF 1.15 too.

image
 

It appears to me, looking through the code that the condition around the iterator reset for the validation set should be the same as the one for the training data set further down. Though I am not 100% sure if this is this.

Script for reproducing the issue:
```

import tensorflow as tf
import tensorflow.keras.backend as K
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Conv2D, Flatten, Dense
import tensorflow.keras.losses as losses
from tensorflow.keras.optimizers import Adam
import numpy as np

def CreateModel():
  model = Sequential()
  model.add(Conv2D(filters=32, kernel_size=3, name='images', activation='relu', kernel_initializer='glorot_normal', input_shape=(39, 39, 3)))
  model.add(Flatten())
  model.add(Dense(10, bias_initializer='zeros', kernel_initializer='glorot_normal'))

  return model

def generatorFunction():
  while 1:
    s = (39, 39, 3)
    z = np.zeros(s, dtype=np.float64)

    zo = np.zeros(10, dtype=np.float64)
    yield z, zo

def RunScript():
  # Setup session options
  config = tf.ConfigProto()

  if tf.test.is_gpu_available():
    config.gpu_options.allow_growth = True

  session = tf.Session(config=config)
  session.run(tf.global_variables_initializer())
  K.set_session(session)

  # Setup strategy
  cross_device_ops = tf.distribute.ReductionToOneDevice(reduce_to_device='/device:CPU:0')
  mirrored_strategy = tf.distribute.MirroredStrategy(cross_device_ops=cross_device_ops)

  # Setup data
  trainingDataPrepped = tf.data.Dataset.from_generator(
    generatorFunction,
    output_types=(tf.float32, tf.float32),
    output_shapes=((39,39,3), (10))).batch(32)

  vDataPrepped = tf.data.Dataset.from_generator(
    generatorFunction,
    output_types=(tf.float32, tf.float32),
    output_shapes=((39,39,3), (10))).batch(32)

  # Create and train model
  with mirrored_strategy.scope():
    model = CreateModel()
    lossType = losses.mean_squared_error
    model.compile(optimizer=Adam(), loss=lossType)


    model.fit(trainingDataPrepped,
              validation_data=vDataPrepped,
              validation_steps=1,
              epochs=3000,
              shuffle=True,
              steps_per_epoch=1)

if __name__ == "__main__":
  RunScript()

```

### CODE REVIEW

**Comment by Loca123** [comment]
@googlebot I signed it!

**Comment by Loca123** [comment]
Note: I don't think this fully explains the slowdown that I am seeing, but I still think that the code fix in this PR is correct.

**Review by omalleyt12 — APPROVED** [review]
Thanks for the PR!

**Comment by Loca123** [comment]
> @Loca123 Could you please check failed build errors? Thanks!

I can see why a validation loss might change, might require an update to the test case. I'm not sure why the Android one failed though, it said it was a 404 error getting data. I just synced my branch to the master but it doesn't seem to be running the same tests that it was doing before.

**Comment by gbaned** [comment]
@omalleyt12  Can you please take a look on this PR? Thanks!

**Comment by omalleyt12** [comment]
Looking at this again, shouldn't the validation iterator be reset every epoch? The validation data should always be fully consumed (as opposed to the training data, which can be partially consumed each epoch)

**Comment by gbaned** [comment]
@Loca123 Could you please check reviewer comments and keep us posted. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by Loca123** [comment]
@gbaned @omalleyt12 : You may very well be right regarding it needing to be reset. Why this leads to a progressive slowdown I do not know. 

However this ultimately did not resolve my original slowdown that is shown in the second graph above. I was only able to resolve that by moving to TF2, there was something in TF1.11 and greater that was causing a severe progressive slowdown, but I could not determine where it was coming from, beyond it being related to the checkpoint callback. Even a very simple script demonstrated it. I believe the above script with a checkpoint callback added would show it.

As for this PR, it might be as well to close it as overall I don't know if the proposed change is correct and has only a little effect on the original reported problem.

Regards,

Carl

**Comment by gbaned** [comment]
@omalleyt12 Can you please take a look on this PR? Thanks!

**Review by omalleyt12 — COMMENTED** [review]
@Loca123 Thanks for the PR!

I think the training logic has now changed enough that this should no longer be an issue. Can you please try this in the latest tf-nightly? `pip install -U tf-nightly`

**Comment by gbaned** [comment]
@Loca123 Could you please check reviewer comments and keep us posted. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
